File: frontend/src/api/survey.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { http } from '../lib/apiClient'

/** Вид вопроса — канон живёт в бэкенде (app/services/survey_form.py). */
export type SurveyQuestionKind = 'text' | 'multi'

export interface SurveyOption {
  key: string
  label: string
}

export interface SurveyQuestion {
  key: string
  kind: SurveyQuestionKind
  title: string
  required: boolean
  hint: string | null
  placeholder: string | null
  options: SurveyOption[]
  min_length: number
  max_length: number
  comment_title: string | null
  comment_required: boolean
}

export interface SurveyForm {
  version: number
  title: string
  subtitle: string
  intro: string
  consent_label: string
  questions: SurveyQuestion[]
  completed_at: string | null
  required: boolean
  gift_available: boolean
}

/** Ответ на один вопрос. Форма зависит от kind — см. survey_form.py. */
export interface SurveyAnswer {
  /** text */
  text?: string
  /** multi */
  choices?: string[]
  comment?: string
}

export type SurveyAnswers = Record<string, SurveyAnswer>

export interface SurveyGift {
  url: string
  expires_in: number
  filename: string
}

export interface SurveyRow {
  user_id: number
  username: string
  display_name: string
  invited: boolean
  completed_at: string | null
  publish_consent: boolean
  has_gift: boolean
  gift_asset_id: number | null
  answers: SurveyAnswers | null
  version: number | null
}

export interface SurveyOverview {
  form: Omit<SurveyForm, 'completed_at' | 'required' | 'gift_available'>
  rows: SurveyRow[]
  invited_count: number
  completed_count: number
}

export const surveyFormKey = ['survey', 'form'] as const
export const adminSurveyKey = ['admin', 'survey'] as const

/** Форма анкеты + состояние текущего пользователя по ней. */
export function useSurveyForm() {
  return useQuery({
    queryKey: surveyFormKey,
    queryFn: () => http.get<SurveyForm>('/api/survey/me'),
    // Не Infinity: книгу админ может привязать уже ПОСЛЕ того, как человек сдал
    // анкету — профиль должен подхватить её без перезагрузки страницы.
    staleTime: 60_000,
  })
}

/** Отправить анкету. После успеха платформа разблокируется (refreshMe). */
export function useSubmitSurvey() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (body: { answers: SurveyAnswers; publish_consent: boolean }) =>
      http.post<{ gift_available: boolean }>('/api/survey', body),
    onSuccess: () => qc.invalidateQueries({ queryKey: surveyFormKey }),
  })
}

/** Presigned-ссылка на личную книгу. Дёргаем по клику, а не заранее. */
export function useSurveyGift() {
  return useMutation({
    mutationFn: () => http.get<SurveyGift>('/api/survey/gift'),
  })
}

/** Админская сводка: кому показана анкета, кто сдал, что ответил. */
export function useAdminSurvey() {
  return useQuery({
    queryKey: adminSurveyKey,
    queryFn: () => http.get<SurveyOverview>('/api/admin/survey'),
  })
}

/** Показать анкету выбранным участникам. */
export function useInviteSurvey() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (userIds: number[]) =>
      http.post<null>('/api/admin/survey/invite', { user_ids: userIds }),
    onSuccess: () => qc.invalidateQueries({ queryKey: adminSurveyKey }),
  })
}

/** Снять блокировку с человека, не дожидаясь анкеты. */
export function useCancelSurveyInvite() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (userId: number) => http.del<null>(`/api/admin/survey/invite/${userId}`),
    onSuccess: () => qc.invalidateQueries({ queryKey: adminSurveyKey }),
  })
}

/** Привязать/отвязать личную книгу участника. */
export function useSetSurveyGift() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: ({ userId, assetId }: { userId: number; assetId: number | null }) =>
      http.patch<null>(`/api/admin/survey/gift/${userId}`, { media_asset_id: assetId }),
    onSuccess: () => qc.invalidateQueries({ queryKey: adminSurveyKey }),
  })
}

File: backend/app/schemas/survey.py
"""Pydantic-схемы выпускной анкеты.

Сами вопросы описаны в `app.services.survey_form` — сюда они попадают уже
сериализованными (`question_form()`), поэтому форма отдаётся `dict`, а не типом
на каждый вид вопроса: канон меняется чаще, чем контракт эндпоинта.
"""
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class SurveyFormOut(BaseModel):
    """Что показывать пользователю: форма + его состояние по ней."""

    version: int
    title: str
    subtitle: str
    intro: str
    consent_label: str
    questions: list[dict[str, Any]]
    # Заполнена ли анкета и ждёт ли она ответа прямо сейчас.
    completed_at: datetime | None
    required: bool
    # Персональная книга привязана — после отправки будет что скачать.
    gift_available: bool


class SurveySubmit(BaseModel):
    """Ответы одним словарём `{ключ вопроса: ответ}` — разбирает survey_form."""

    model_config = ConfigDict(extra="forbid")

    answers: dict[str, Any]
    publish_consent: bool = False


class SurveyGiftOut(BaseModel):
    """Presigned-ссылка на личную книгу."""

    url: str
    expires_in: int
    filename: str


class SurveyRowOut(BaseModel):
    """Строка админской таблицы: один участник."""

    user_id: int
    username: str
    display_name: str
    invited: bool
    completed_at: datetime | None
    publish_consent: bool
    has_gift: bool
    gift_asset_id: int | None
    answers: dict[str, Any] | None
    version: int | None


class SurveyOverviewOut(BaseModel):
    """Админская сводка: форма (для подписей) + строки + счётчики."""

    form: dict[str, Any]
    rows: list[SurveyRowOut]
    invited_count: int
    completed_count: int


class SurveyInviteRequest(BaseModel):
    """Кому показать анкету. Пустой список — ошибка, нечего делать."""

    model_config = ConfigDict(extra="forbid")

    user_ids: list[int] = Field(min_length=1)


class SurveyGiftRequest(BaseModel):
    """Привязать книгу к участнику (или отвязать: media_asset_id = null)."""

    model_config = ConfigDict(extra="forbid")

    media_asset_id: int | None
